docs/proposals/local-cluster-ux.md

```diff
@@ -0,0 +1,184 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubernetes Local Cluster Experience
+
+This proposal attempts to improve the existing local cluster experience for kubernetes.
+The current local cluster experience is sub-par and often not functional.
+This proposal is a RFC!
+
+## Primary Goals
+
+From a high level the goal is to make it easy for a new user to run a Kubernetes cluster and play with curated examples that require least amount of knowledge about Kubernetes.
+These examples will only use kubectl and only a subset of Kubernetes features that are available will be exposed.
+
+- Works across multiple OSes - OS X, Linux and Windows primarily.
+- Single command setup and teardown UX.
+- Unified UX across OSes
+- Minimal dependencies on third party software.
+- Minimal resource overhead.
+- Eliminate any other alternatives to local cluster deployment.
+
+## Secondary Goals
+
+- Enable developers to use the local cluster for kubernetes development.
+
+## Non Goals
+
+- Simplifying kubernetes production deployment experience
+- Supporting all possible deployment configurations of Kubernetes like various types of storage, networking, etc.
+
+
+## Local cluster requirements
+
+- Includes all the master components & DNS (Apiserver, scheduler, controller manager, etcd and kube dns)
+- Basic auth
+- Service accounts should be setup
+- Kubectl should be auto-configured to use the local cluster
+- Tested & maintained as part of Kubernetes core
+
+## Existing solutions
+
+Following are some of the existing solutions that attempt to simplify local cluster deployments.
+
+### [Spread](https://github.com/redspread/spread)
+
+Spread's UX is great!
+It is adapted from monokube and includes DNS as well.
+It satisfies almost all the requirements, excepting that of requiring docker to be pre-installed.
+It has a loose dependency on docker.
+New releases of docker might break this setup.
+
+### [Kmachine](https://github.com/skippbox/kmachine)
+
+Kmachine is adapted from docker-machine.
+It exposes the entire docker-machine CLI.
+It is possible to repurpose Kmachine to meet all our requirements.
+
```

Comment:
On Wed, Apr 13, 2016 at 01:01:12AM -0700, runseb wrote:
> > +Following are some of the existing solutions that attempt to simplify local cluster deployments.
> > +
> > +### [Kmachine](https://github.com/skippbox/kmachine)
> > +
> > +Kmachine is adapted from docker-machine.
> > +It exposes the entire docker-machine CLI.
> > +It is possible to repurpose Kmachine to meet all our requirements.
> > +
> 
> I very often have multiple 'docker-machine' running in different clouds. One bonus of doing it is that the bandwidth to the cloud provider is usually much higher than on your local setup, so images download much faster.
> 
> So speaking for myself, I tend to develop on a droplet in digitalocean rather than on a local virtual box.
> 
> So say, you run localkube in a VM on vbox (if I understand this proposal correctly), then I bet you that someone is going to ask to run this on cloud providers, and you will go down the route of docker-machine like drivers.

I'm not sure this will happen, as this is a "toy" example and there are ways to
run on cloud providers and are quite easy. You can "play" locally and when you
want to try more things, you deploy to a cloud provider and test what you want
(load balancers, PV, etc.).

It's not clear to me that this is/will be needed.

> Also when doing something purely local you tend to make assumptions that are not valid remotely, and when you move to a cloud your stuff breaks.

Like which ones? It didn't happen to me. YAML files that are valid in a local
kubernetes, are valid on a remote one. Just that locally you have less features
(like PVs or LB), but that is just the reason I don't expect this to happen.

> I will also note that kubectl is made to handle multiple remote endpoints and that in the production case your cluster will be remote, so  I believe it is better to build a single node experience that still mimics remote k8s access.

The toolbox proposes to use kubectl, I understand.